### PR TITLE
Add support on the dataflow CFG for visiting lambda result expressions

### PR DESCRIPTION
**Pull request by lazaroclapp**
This adds a new node to the cfg, LambdaResultExpressionNode, which corresponds
to the result of a lambda expression with a single expression as its body, and
which is added at the end of the cfg when CFGBuilder is passed a CFGLambda of
this kind as the underlaying ast.

This is needed by some analyses to handle the result of a lambda expression in
the case where the body is a single ExpressionTree. The case where the body is
a block is already handled by matching ReturnNode.

### CODE REVIEW

**Comment by lazaroclapp** [comment]
I am also not sure if there is any other code internally in the checkerframework that needs to be updated to treat `LambdaResultExpressionNode` in ways similar to how `ReturnNode` is currently handled.

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * @author Lazaro Clapp
+ */
+public class LambdaResultExpressionNode extends Node {
+
+    protected ExpressionTree tree;
+    protected /*@Nullable*/ Node result;
+
+    public LambdaResultExpressionNode(
+            ExpressionTree t, /*@Nullable*/ Node result, Types types /*,
+            LambdaExpressionTree lambda,
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Can these comments be removed?

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+*/
+
+/**
+ * A node for the single expression body of a single expression lambda
+ *
+ * @author Lazaro Clapp
+ */
+public class LambdaResultExpressionNode extends Node {
+
+    protected ExpressionTree tree;
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Can you add some documentation what the tree and result represent, either here or, maybe better, with the public getter methods?

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    protected /*@Nullable*/ Node result;
+
+    public LambdaResultExpressionNode(
+            ExpressionTree t, /*@Nullable*/ Node result, Types types /*,
+            LambdaExpressionTree lambda,
+            Symbol.MethodSymbol methodSymbol*/) {
+        super(InternalUtils.typeOf(t));
+        this.result = result;
+        tree = t;
+        //result.setAssignmentContext(new AssignmentContext.LambdaReturnContext(methodSymbol));
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Did you run into any problems with the assignment context? Does this require more work or can this comment be removed?

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || !(obj instanceof LambdaResultExpressionNode)) {
+            return false;
+        }
+        LambdaResultExpressionNode other = (LambdaResultExpressionNode) obj;
+        if ((result == null) != (other.result == null)) {
+            return false;
+        }
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Can you add a comment why the `tree` doesn't need to be compared?

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        return (result == null || result.equals(other.result));
+    }
+
+    @Override
+    public int hashCode() {
+        return HashCodeUtils.hash(result);
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Similarly, why does the hash only depend on the `result`, not on the `tree`? Maybe this is something to document with the fields, also.

**Comment by wmdietl** [comment]
Regarding other code changes: I tested your changes a bit and everything seemed to work as expected - we already had test coverage for lambdas with single expressions.
I opened one issue for a possible follow-up, #1607, but that shouldn't block this PR.

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java (hunk 2 above)
Comment (lazaroclapp) [reply dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Adding it to the methods, let me know if its clear enough (working on the code as I reply to the comments, will update the PR shortly).

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java (hunk 3 above)
Comment (lazaroclapp) [reply dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Nope. Removing, this is left-over code from another approach at this.

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java (hunk 4 above)
Comment (lazaroclapp) [reply dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: You know what, let me add a comparison for `tree` too. It shouldn't matter if we are constructing `LambdaResultExpressionNode`'s correctly always (`result` is supposed to be `scan([path_to_tree], null);`, but we are not enforcing this within the class, so lets both check here and use it to compute the hash)

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java (hunk 5 above)
Comment (lazaroclapp) [reply dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: See above.

File: dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java (hunk 4 above)
Comment (lazaroclapp) [reply dataflow/src/org/checkerframework/dataflow/cfg/node/LambdaResultExpressionNode.java]: Although, actually, `ReturnNode`, which is what I am basing this class on, doesn't use `tree` as part of its comparison... mmm...

**Comment by lazaroclapp** [comment]
Rebased and addressed the above comments. Thanks!! :)

**Pull request merged**